### ISSUE TITLE
over fittingを避けるためのヒント

### ISSUE THREAD

**Issue by riron1206** [riron1206/read_paper#29]
https://www.kaggle.com/c/lish-moa/discussion/196913
■over fitting を避けるには
同じvalidation dateを頻繁に使用しないようにする
（同じデータで精度評価を繰り返すと絶対にover fitting する）
over fitting 回避策
1. k-fold cv + public LBを追加のfoldとして利用
（これでもover fitting するが、1つだけのvalidation setよりまし）
2. k-fold cv のfold の数を時々変更して評価
（計算コスト高いので難しい）
※2段階を超えるスタッキングはover fitting を起こす。多段スタッキングはだめ

■モデルが精度向上したといえるには
1. k-fold cv スコアが十分に改善
2. すべてのフォールドスコアが十分に改善
※0.00005または0.0001のような小さなスコアの改善は意味がない
※ノイズの効果を評価する場合は、同じフォールドで異なるランダムシードを使用してモデルを数回トレーニングし、スコアの標準偏差と平均を確認する
→単純なモデルから始めて、CVが大幅に増加した場合にのみ変更を加えるのが良い

■foldは増やした方がいいのか
一般的には、より多くのモデルをトレーニングし、各モデルはより多くのデータでトレーニングされるため、cvとLBがと十分に相関していればスコア上がる
ただし、増やせば増やすほど良いわけではない
foldの数を増やすと、各foldのモデルが似てくるため、すべてのモデルをブレンドしてもパフォーマンスが常に向上するとは限らない

■PCAなどのデータを元にした特徴量について
PCAに合わせるために何を使用するかによって異なる
1. train setだけでfitしたPCAの特徴量の場合
　外部データからのリークなし

2. train set + validation setでfitしたPCAの特徴量の場合
　少しオーバーフィットする。CVが下がり、LBが上がるはず

3. train set + validation set + public LB でfitしたPCAの特徴量の場合
　CVとpublic LBの両方が低下することがあるが、private LBが上昇する可能性ある

4. train set + validation set + public LB + private LB でfitしたPCAの特徴量の場合
　CV、public LB + private LBが低下する場合があるが、新しいデータではスコアが上昇します
　Kaggleでより適切に機能する可能性がある

■関連論文
Model Evaluation, Model Selection, and Algorithm Selection in Machine Learning
https://arxiv.org/abs/1811.12808
小さいデータセットでのmodelの評価は 5foldのnested cross-validation を推奨していた
(外側のループで5分割交差検定、内側のループで2分割交差検定)
1. 5foldのtrain/test setに分ける
2. 各foldのtrain setをtrain:val=1:1 で2foldに分割
3. 各fold評価する
サンプルコード: https://github.com/rasbt/model-eval-article-supplementary/blob/master/code/nested_cv_code.ipynb
![fig](https://media.arxiv-vanity.com/render-output/3560955/x22.png)
